RemoteSpiNNaker/RemoteSpiNNakerWeb/src/main/java/uk/ac/manchester/cs/spinnaker/output/OutputManagerImpl.java

```diff
@@ -243,44 +241,33 @@ private File getProjectDirectory(final String projectId) {
     }
 
     @Override
-    public List<DataItem> addOutputs(final String projectId, final int id,
-            final File baseDirectory, final Collection<File> outputs)
-            throws IOException {
-        if (outputs == null) {
-            return null;
-        }
-
+    public DataItem addOutput(final String projectId, final int id,
+            final String baseDirectory, final String output,
+            final InputStream input) throws IOException {
         final String pId = new File(projectId).getName();
-        final int pathStart = baseDirectory.getAbsolutePath().length();
+        final int pathStart = baseDirectory.length();
         final File projectDirectory = getProjectDirectory(projectId);
         final File idDirectory = new File(projectDirectory, String.valueOf(id));
 
         try (JobLock op = new JobLock(idDirectory)) {
-            final List<DataItem> outputData = new ArrayList<>();
-            for (final File output : outputs) {
-                if (!output.getAbsolutePath()
-                        .startsWith(baseDirectory.getAbsolutePath())) {
-                    throw new IOException("Output file " + output
-                            + " is outside base directory " + baseDirectory);
-                }
-
-                String outputPath = output.getAbsolutePath()
-                        .substring(pathStart).replace('\\', '/');
-                if (outputPath.startsWith("/")) {
-                    outputPath = outputPath.substring(1);
-                }
+            if (!output.startsWith(baseDirectory)) {
```

Comment:
You also need a check if the name contains `..` as a component.

And you ought to do the name sanity checks before acquiring the lock; they don't (and shouldn't) depend on the state of the named entities.